src/main/java/appeng/tile/inventory/AppEngInternalAEInventory.java

```diff
@@ -150,42 +141,85 @@ public ItemStack getStackInSlot( final int var1 )
 	}
 
 	@Override
-	public ItemStack decrStackSize( final int slot, final int qty )
+	public ItemStack insertItem( int slot, ItemStack stack, boolean simulate ) 
+	{
+		 if ( stack.isEmpty() )
+		 {
+	            return ItemStack.EMPTY;
+		 }
+
+		 ItemStack existing = getStackInSlot( slot );
+		 int limit = getStackLimit( slot, stack );
+
+		 if ( !existing.isEmpty() )
+	     {
+			 if ( !ItemHandlerHelper.canItemStacksStack( stack, existing ) )
+				 return stack;
+
+			 limit -= existing.getCount();
+	     }
+
+		 if ( limit <= 0 )
+			 return stack;
+
+		 boolean reachedLimit = stack.getCount() > limit;
+
+	     if ( !simulate )
+	     {
+	    	 if ( existing.isEmpty() )
+	         {
+	    		 this.inv[slot] = AEApi.instance().storage().createItemStack( reachedLimit ? ItemHandlerHelper.copyStackWithSize( stack, limit ) : stack );
+	         }
+	         else
+	         {
+	        	 existing.grow( reachedLimit ? limit : stack.getCount() );
+	         }
+	    	 if( this.te != null && Platform.isServer() )
+	    	 {
+				this.te.onChangeInventory( this, slot, InvOperation.insert, ItemStack.EMPTY,  reachedLimit ? ItemHandlerHelper.copyStackWithSize( stack, limit) : stack );
+	    	 }	    	 
+	     }
+	     return reachedLimit ? ItemHandlerHelper.copyStackWithSize( stack, stack.getCount()- limit ) : ItemStack.EMPTY;
+	}
+
+	@Override
+	public ItemStack extractItem( int slot, int amount, boolean simulate ) 
 	{
 		if( this.inv[slot] != null )
-		{
+		{		
 			final ItemStack split = this.getStackInSlot( slot );
-			ItemStack ns = ItemStack.EMPTY;
 
-			if( qty >= split.getCount() )
+			if( amount >= split.getCount() )
 			{
-				ns = this.getStackInSlot( slot );
-				this.inv[slot] = null;
+				if ( !simulate )
+				{
+					this.inv[slot] = null;
+					if( this.te != null && Platform.isServer() )
+					{
+						this.te.onChangeInventory( this, slot, InvOperation.extract, split, ItemStack.EMPTY );
+					}
+				}
+				return split;
 			}
 			else
 			{
-				ns = split.splitStack( qty );
-			}
-
-			if( this.te != null && Platform.isServer() )
-			{
-				this.te.onChangeInventory( this, slot, InvOperation.decreaseStackSize, ns, ItemStack.EMPTY );
+				if ( !simulate )
+				{
+					split.grow( -amount );
+					if( this.te != null && Platform.isServer() )
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS134)